lib/kafka/connection.rb

```diff
@@ -172,23 +230,30 @@ def read_response(response_class, notification)
 
       return correlation_id, response
     rescue Errno::ETIMEDOUT
-      @logger.error "Timed out while waiting for response #{@correlation_id}"
+      @logger.error "Timed out while waiting for response #{expected_correlation_id}"
       raise
     end
 
-    def wait_for_response(response_class, notification)
+    def wait_for_response(response_class, expected_correlation_id, notification)
       loop do
-        correlation_id, response = read_response(response_class, notification)
-
-        # There may have been a previous request that timed out before the client
-        # was able to read the response. In that case, the response will still be
-        # sitting in the socket waiting to be read. If the response we just read
-        # was to a previous request, we can safely skip it.
-        if correlation_id < @correlation_id
-          @logger.error "Received out-of-order response id #{correlation_id}, was expecting #{@correlation_id}"
-        elsif correlation_id > @correlation_id
-          raise Kafka::Error, "Correlation id mismatch: expected #{@correlation_id} but got #{correlation_id}"
+        correlation_id, response = read_response(response_class, expected_correlation_id, notification)
+
+        if correlation_id < expected_correlation_id
+          # There may have been a previous request that timed out before the client
+          # was able to read the response. In that case, the response will still be
+          # sitting in the socket waiting to be read. If the response we just read
+          # was to a previous request, we deliver it to the pending async response
+          # future.
+          async_response = @pending_async_responses.delete(correlation_id)
+          async_response.deliver(response) if async_response
```

Comment:
hmmm, you can pipeline requests, but you have to block on them in the order you sent them in; maybe there's a bit deeper on an abstraction here where you send multiple requests and then get back a single handle that will resolve each request in turn?  Something like:

```
batch = broker.start_batch
topics_by_broker.each { |t|
   batch.fetch_messages_async(...)
}
batch.each_response { |response|

}
```

just  a thought.